### DISCUSSION TITLE
Possible to parse expressions between delimiters inside of math environments?

### DISCUSSION THREAD

**Discussion by mattvague** [micromark/micromark#111] [Q&A] [answered]
Hey there, this is kind of a continuation of the question I asked [here](https://github.com/mdx-js/mdx/discussions/1991). I've made a fair bit of progress on this to the point where I can now transform e.g. ``What is $1 + 1$`` ➡️  ``What is <Katex>{`1 + 1`}</Katex>``, I'll try and post how I accomplished as a follow up answer once I'm done.

The thing I need to figure out now though is how to map something like `What is $<a> + <b>$`  ➡️ ``What is <Katex>{`${a} + ${b}`}</Katex>`` where `<` and `>` are arbitrary delimiters.

**My Question**: is it possible to write a micromark extension that works _with_ [micromark-extension-math](https://github.com/micromark/micromark-extension-math) in such a way that for a given math environment I can tokenize any contained expressions first, then let micromark-extension-math do it's own parsing? I'm not familiar enough with the ecosystem to know if this is even possible without forking micromark-extension-math so would be very grateful if someone can point me in the right direction, here's what I have so far:

```
import { ok as assert } from "uvu/assert";
import { codes, values } from "micromark-util-symbol/codes.js";

export function cas() {
  return {
    text: {
      [codes.dollarSign]: mathWithExpressionsText(options)
    },
  };
}

export const mathWithExpressionsText = (options) => {
  const { expressionStartDelimiter = values.lessThan, expressionEndDelimiter = codes.greaterThan } = options;

  return {
    name: "cas",
    tokenize: tokenizeCasMath,
    add: 'before'
  };

  function tokenizeCasMath(effects, ok, nok) {
    return start;

    function start(code) {
      assert(code === codes.dollarSign, `expected \`${codes.dollarSign}\``);
      console.log("tokenizeCasMath", { code });
      return insideMath(code);
    }

    function insideMath(code) {
      console.log("insideMath", { code });

      if (code === expressionStartDelimiter) {
        // Do my custom expression parsing
      }

      // Somehow pass the parsing back to micromark-math-extension?
    }
  }
};
```

Thanks a million!

Matt

**Comment by wooorm** [comment, marked as answer]
Heya!

> e.g. What is `$1 + 1$` ➡️ What is `<Katex>{`1 + 1`}</Katex>`, I'll try and post how I accomplished as a follow up answer once I'm done.

I’m not sure what you did, but this is very close to what `remark-math` already exposes, plus a small tree transform afterwards? Is that what you’re doing (and if not, why are you doing more complicated things?)

> how to map something like What is `$<a> + <b>$` ➡️ What is `<Katex>{`${a} + ${b}`}</Katex>` where `<` and `>` are arbitrary delimiters.

It would help a lot here if I know what your **actual input** is, and your **expected output**. What is the syntax of your variables? What are the limitations? Is it JavaScript? Can they include complete programs? Statements? Are they just identifiers?

> My Question

No, it is not possible to add custom parsing to `micromark-extension-math`, or practically to “hijack” the occurrence of `{` when it is found in math
Depending on the previous question, you might not need to.
Alternatively, your attached code shows a start of a fork, which is indeed another way to go about it.

**Reply by mattvague** [reply]
>I’m not sure what you did, but this is very close to what remark-math already exposes, plus a small tree transform afterwards? Is that what you’re doing (and if not, why are you doing more complicated things?)

Yep, I am using remark-math for parsing but instead of rendering katex at compile time with `rehype-katex`, I have a custom plugin transforming math nodes to a custom `katex` HTML tag. This tag is then mapped by MDX to my `<Katex>` component which renders katex at runtime, making it possible to support dynamic values in katex expressions as shown.

>It would help a lot here if I know what your actual input is, and your expected output. What is the syntax of your variables? What are the limitations? Is it JavaScript? Can they include complete programs? Statements? Are they just identifiers?

I'm hoping to be able to support any valid JS expression that can be stringified, with full access to variables exported in the rest of the MDX doc.

E.g.

```
$2\pi = <2 * Math.PI>$
```

or

```
export const myVar = props.myVar;

$$
  \f\relax{<myVar>} = <myVar>^2 - <1 * 2 + 3>
$$
```

> No, it is not possible to add custom parsing to micromark-extension-math, or practically to “hijack” the occurrence of { when it is found in math

Dang, that's what I figured 😔 By "{" are you meaning whatever arbitrary delimiter I'm choosing to start my expression? Is there any way I could transform the math tokens outputted by `micromark-extension-math` after the fact or something?

Thanks again for your help! I know this is quite a non-standard use case, but unfortunately my project demands it. MDX and the unified ecosystem are such a perfect fit for my project otherwise that I think it makes sense for me to put in whatever effort is necessary here.

**Reply by mattvague** [reply]
Took at walk, and think I had a simpler idea. Katex allows defining custom macros implemented with a JS function, so maybe rather than introducing some new delimiter syntax in math envs, I wonder if I could just have a custom macro that allowed evaluating & stringifying these expressions (off without opening up a giant security hole), and somehow also have access to the exports in the current MDX document

Example

```
export const myVar = props.myVar;

$$
  \f\relax{\eval{myVar}} = \eval{myVar}^2 - \eval{1 * 2 + 3}
$$
```

I think the second part will be trickier as I'm not sure how to get access to the current exports from the MDX document, will post a followup if/when I figure it out

**Reply by wooorm** [reply]
> custom plugin transforming math nodes to a custom `katex` HTML tag, This tag is then mapped by MDX to my `<Katex>` component

Perhaps of interest: you can create nodes for JSX too: https://mdxjs.com/packages/remark-mdx/#syntax-tree.

> I'm hoping to be able to support any valid JS [...]

And should those expressions be marked-up by KaTeX? Or should they be replaced with what the expressions evaluate to before being passed to KaTeX?

> By `"{"` are you meaning whatever arbitrary delimiter 

Yep. In short: micromark is a finite state machine. That means that when `$` is found, it goes into some process that only that process can exit, because all the states are defined. There are exceptions with that extensions can define say `$` to now start math in a certain content type (e.g., flow/block, text/inline).

> Is there any way I could transform the math tokens outputted by `micromark-extension-math` after the fact or something?

You’d have problem with `$` in strings, comments, or just anywhere in JavaScript, e.g., `$2\pi = <2 /* $ */ * Math.PI>$`.
It’s also not what tokens are for. That’s what ASTs are for. Which is what I alluded to with “plus a small tree transform afterwards”.

**Reply by mattvague** [reply]
> And should those expressions be marked-up by KaTeX? Or should they be replaced with what the expressions evaluate to before being passed to KaTeX?

Katex should receive a string where all expressions have been replaced with their evaluated value, basically the same as ``<Katex>{`${a} + ${b}`}</Katex>`` where my component's children prop is string with whatever the values of `a` and `b` evaluated to.

>You’d have problem with $ in strings, comments, or just anywhere in JavaScript, e.g., $2\pi = <2 /* $ */ * Math.PI>$.
It’s also not what tokens are for. That’s what ASTs are for. Which is what I alluded to with “plus a small tree transform afterwards”.

Right sorry, that makes more sense. I guess I'm not totally sure how that transform would work though. Are you meaning that my plugin would visit any `math` and `inlineMath` nodes, and transform them something like this:

```js
// Before
{
  type: 'inlineMath',
  value: '<a> + <b>',
  data: {
    hName: 'span',
    hProperties: {className: ['math', 'math-inline']},
    hChildren: [{type: 'text', value: '<a> + <b>'}]
  }
}

// After
{
  type: 'mdxJsxFlowElement',
  name: 'Katex',
  attributes: [],
  children: [
    {
      type: 'mdxTextExpression',
      value: '`${a} + ${b}`'
    },
  ]
}
```

If so, what's the best way to parse the `value` of a node like this into a subtree?

**Reply by mattvague** [reply]
By the way, I have to say it's mind blowing how many packages you maintain 🤯 I'd like to support your work in some way, do you (or any of your projects) have a way to accept small financial contributions?

**Reply by wooorm** [reply]
How about this. Given your proposal MDX:

```markdown
export const myVar = props.myVar;

$$
  \f\relax{\eval{myVar}} = \eval{myVar}^2 - \eval{1 * 2 + 3}
$$
```

…and given the variable-use syntax which could be `{myVar}` (above) or `&myVar` or whatever.
Which we want to turn into (I believe the escapes are correct but treat it as pseudo-code) the following MDX/JSX:

```js
export const myVar = props.myVar;

<Katex>{`\\f\\relax{\\eval${myVar}} = \\eval${myVar}^2 - \\eval{1 * 2 + 3}`}</Katex>
```

JSX compiled away, it would be something like (pseudo-code):

```js
export const myVar = props.myVar;

React.createElement(Katex, {
  children: `\\f\\relax{\\eval${myVar}} = \\eval${myVar}^2 - \\eval{1 * 2 + 3}`
})
```

That can be achieved with:

- use `remark-math` to parse it as math in markdown as normal
- make a new remark plugin that visits `math`, `inlineMath` nodes
- which turns them to `mdxJsxFlowElement`, `mdxJsxTextElement`: https://github.com/syntax-tree/mdast-util-mdx-jsx#syntax-tree
- paste the “goal mdx/jsx” into the [playground](https://mdxjs.com/playground/) to see what AST you need to create
- one big Q that’s left is how to interpret the variables. You could treat everything inside the math as JS, but then all the escapes need to be double escaped, which gets really ugly and hard to type. Perhaps it’s easier to find-and-replace on `/{\w+}/g`? Or maybe `/\$\w+/g`?

---

I see you’ve found the button, thanks for sponsoring! 💖
To still answer here: I do maintain a lot! Most of it in unified, which is larger than me (see https://github.com/unifiedjs/collective), and some of it is “just me”. For the links, see https://wooorm.com/thanks/.

**Reply by mattvague** [reply]
Yay! Finally got it all working. Thanks a ton for the playground tip, would have taken me a lot longer to figure out what shape of tree I needed to produce otherwise. 

@wooorm If interested, here's what I ended up with. Can you see any big issues? Might it be worth releasing as a plugin?

```js
import { visit, CONTINUE } from 'unist-util-visit'
import { Parser } from 'acorn'

/**
 * Plugin to transform math nodes (like from remark-math) to JSX element nodes which render math at run time (likely using Katex)
 * 
 * Supports JS expressions inside of math similar to how MDX supports JS expressions inside of {...}
 * Allows for dynamic math expressions not possible with usage of something like rehype-katex
 *
 * e.g.
 * 
 * export const pi = Math.PI
 * 
 * $\pi = \js{pi}$
 * 
 * is transformed to
 * 
 * <Katex>{String.raw`\pi = ${pi}`}</Katex>
 * 
 * **Note** this plugin expects math to be rendered at run time inside of a React compoent instead of
 * at compile time like rehype-katex. This means user's browsers have to do more work and should be used
 * only when dynamic math (i.e. math with JS expressions inside) is required.
 * 
 * @type {import('unified').Plugin}
 * @param {Object} options
 * @param {string} options.component - Name of react component to transform remark math nodes to (which will render math)
 * @param {RegExp} options.expressionPattern - Regex to match JS expressions inside of math to convert to `${...}`. Default is to match `\js{...}`
 */
export default function remarkMathMdxPlugin(
  options = {
    component: 'Katex',
    expressionPattern: /\\js\{([^\{\}]+)\}/gm
  }
) {
  return (tree) => {
    visit(tree, 'math', (node, index, parent) => {
      parent.children.splice(index, 1, {
        type: 'mdxJsxFlowElement',
        name: options.component,
        attributes: [
          {
            type: 'mdxJsxAttribute',
            name: 'display'
          }
        ],
        children: [{
          type: 'mdxFlowExpression',
          value: node.value,
          data: {
            estree: transformMathNodeToEstree(node)
          }
        }]
      })
      return [CONTINUE, index]
    })

    visit(tree, 'inlineMath', (node, index, parent) => {
      parent.children.splice(index, 1, {
        type: 'mdxJsxTextElement',
        name: options.component,
        attributes: [],
        children: [{
          type: 'mdxTextExpression',
          value: node.value,
          data: {
            estree: transformMathNodeToEstree(node)
          }
        }]
      })
      return [CONTINUE, index]
    })
  }

  /**
   * Replaces any instances of `\js{...}` in a string with `${...}` in order to 
   * evaluate the contents of the macro as javascript
   * 
   * @param {string} value - String to transform
   */
   function transformJSExpressions(string) {
    return string.replace(options.expressionPattern, '${$1}').trim()
  }

  /**
   * Parse the the contents of a Math node as a template string to ESTree 
   * 
   * @param {import('unist').Node} node - Node to transform
   * @returns {import('acorn').Node} node - Estree resulting from parsing math node contents
   */
  function transformMathNodeToEstree(node) {
    return Parser.parse(`String.raw\`${transformJSExpressions(node.value)}\``)
  }
}

```

**Reply by wooorm** [reply]
Good to hear! It looks really good. Some thoughts reading through your code:

* the code for `'math'` and `'inlineMath'` is very similar, you can probably better do one `visit` (on every node), then check if the node is a math node, and then use if-statements for the differing parts?
* instead of children, you could pass it as an attribute as well. Not sure which is better
* `String.raw` is smart!
* you should probably set `sourceType: 'module'` (I think that that type is exposed on the `Program` node, and in MDX it should be a module instead of a script), but otherwise I see no need to pass options to acorn in this case
* you can add support for math nodes in the types: https://github.com/remarkjs/remark-math/tree/main/packages/remark-math#types, this example also defines the mdast tree as being expected inside the plugin, so it’d type the `tree`

**Reply by mattvague** [reply]
Great feedback, thanks! If I were to pull this into a module, do you think the name `remark-math-mdx` would make sense? Do you think there would be value in doing so, or is this use case too specific?

**Reply by wooorm** [reply]
Two ideas:

- `remark-mdx-math` (\*the\* mdx and math plugin)
- `remark-mdx-math-enhanced` (\*a\* useful mdx and math plugin)

Think the `remark-mdx` prefix makes sense. Then it can also be listed on https://mdxjs.com/docs/extending-mdx/ and https://mdxjs.com/guides/math/

**Reply by mattvague** [reply]
Sweet! How's this for a start? https://github.com/goodproblems/remark-mdx-math-enhanced

`remark-mdx-math-enhanced` sounds like it might be the least confusing compared to `remark-math`?

It's definitely not ready for prime time yet; I think I'll want to test it a bit in my own app before sharing it widely but would be great to get your veteran eyes on it in the mean time though. Thanks!

**Reply by wooorm** [reply]
* I believe the type annotations in `visit<Root>(tree, (node, index, parent: Parent) => {` are not needed?
* You can get the math and mdx nodes “into” the tree with https://github.com/remarkjs/remark-math/tree/main/packages/remark-math#types, https://mdxjs.com/packages/remark-mdx/#types
* then you should be able to inline the `isInline?MathNode`s, as the types support them, and you don’t need to cast to `Literal` but they’ll be the actual math node types
* I don’t think the `return [CONTINUE, index];`s are needed, or any return value. The default is `[CONTINUE, index + 1]` I believe, and there shouldn’t be math nodes in math nodes, so those are equivalent

---

Looks fine to me. Feel free to add them to the lists!

**Reply by mattvague** [reply]
Alright, feedback has been addressed and package is published: https://www.npmjs.com/package/remark-mdx-math-enhanced. Will open a PR to add to the lists. Thanks!
